File: pset2/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

int validateKey(string key);
int getKey(string key, int index);
char shiftValue(char input, char max, int key);

int main(int argc, string argv[])
{
    if(argc != 2)
    {
        printf("Usage: %s k\n", argv[0]);
        return 1;
    }

    string key = argv[1];
    int validation = validateKey(key);
    if(validation != 0)
    {
        return validation;
    }

    printf("plaintext: ");
    string s = get_string();

    printf("ciphertext: ");
    int index = 0;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        int currentKey = getKey(key, index);
        char original = s[i];
        char encrypted = original;

        if ('A' <= original && original <= 'Z')
        {
            encrypted = shiftValue(encrypted, 'Z', currentKey);
            index++;
        }
        else if ('a' <= original && original <= 'z')
        {
            encrypted = shiftValue(encrypted, 'z', currentKey);
            index++;
        }

       printf("%c", encrypted);
    }
    printf("\n");
}

int getKey(string key, int index)
{
    int length = strlen(key);
    char character = key[index % length];

    if ('A' <= character && character <= 'Z')
    {
        return character - 'A';
    }
    else
    {
        return character - 'a';
    }
}

int validateKey(string key)
{
    for (int i = 0, n = strlen(key); i < n; i++)
    {
        char c = key[i];
        if (c < 'A' || c > 'z' || (c > 'Z' && c < 'a'))
        {
            return 1;
        }
    }

    return 0;
}

char shiftValue(char input, char max, int key)
{
    int encrypted = input + key;
    if (encrypted > max)
    {
        encrypted -= 26;
    }

    return (char)encrypted;
}

File: pset7/vigenere/vigenere.py
import sys
from cs50 import get_string

def getKey(key, index):
    length = len(key)
    character = key[index % length]

    if 'A' <= character and character <= 'Z':
        return chr(ord(character) - ord('A'))
    else:
        return chr(ord(character) - ord('a'))

def validateKey(key):
    l = len(key)
    for i in range(l):
        c = key[i]
        if (c < 'A' or c > 'z' or (c > 'Z' and c < 'a')):
            return 1

    return 0

def shiftValue(input, max, key):
    encrypted = ord(input) + ord(key)
    if encrypted > ord(max):
        encrypted -= 26

    return chr(encrypted)

if len(sys.argv) != 2:
    print(f"Usage: {sys.argv[0]}\n")
    sys.exit(1)

key = sys.argv[1]
validation = validateKey(key)
if validation != 0:
    sys.exit(validation)

s = get_string("plaintext: ")
print("ciphertext: ", end="")
index = 0
l = len(s)
for i in range(l):
    currentKey = getKey(key, index)
    original = s[i]
    encrypted = original

    if ('A' <= original and original <= 'Z'):
        encrypted = shiftValue(encrypted, 'Z', currentKey)
        index += 1
    elif ('a' <= original and original <= 'z'):
        encrypted = shiftValue(encrypted, 'z', currentKey)
        index += 1

    print(encrypted, end="")

print("")